### ISSUE TITLE
[Bug] `patch:` and `file:` protocols don't work with relative paths in `packageExtensions`

### ISSUE THREAD

**Issue by goszczynskip** [yarnpkg/berry#1127]
**Describe the bug**

Patching fails with: `ENOENT: no such file or directory, open './foo.patch'` when trying to install ~with `nodeLinker: "node-modules"` set~.

**To Reproduce**

Repo with example: https://github.com/goszczynskip/yarn-berry-bug-patch

Steps that I made:
* created yarn repo
* ~set nodeLinker to node-modules~ EDIT: with pnp it fails too
* patched watchpack from webpack
* add webpack to dependencies
* run `yarn`

**Environment if relevant (please complete the following information):**

 - OS: Windows 10
 - Node version v13.2.0
 - Yarn version 2.0.0-rc.31 and sources

**Comment by paul-soporan** [comment]
(Thanks to @arcanis for explaining this to me first. 😄)

The current behavior is intended. This is kind of an edge case that affects all of the protocols that accept paths (`patch:`, `file:`, `portal:` etc.).

Dependencies declared through the `packageExtensions` field simply replace a package's dependencies, so relative paths are resolved relative to the path of the original package (in this case the root of the zip archive containing the package).
On the other hand, if the [`resolutions`](https://yarnpkg.com/configuration/manifest#resolutions) field contains dependencies with protocols with relative paths, they will be resolved relative to the project (your project's) root.

In short:
- The `resolutions` field should be used to override dependencies already declared (i.e. to apply your custom fixes).
- The `packageExtensions` field should be used to add undeclared dependencies.

I'll add a note about this in the description of the `resolutions` field.
